src/redisai.c

```diff
@@ -114,41 +113,37 @@ size_t deviceToKey(RAI_Device device, int64_t deviceid) {
   return device * (RAI_MAX_DEVICE_ID + 1) + (deviceid + 1);
 }
 
+/* Ensure that the the run queue for the device exists.
+ * If not, create it. As an example, if you ask for queue number 4,
+ * all queues for devices behind it ( 0,1,2,3 ) will be created if they don't exist. */
 int ensureRunQueue(RAI_Device device, int64_t deviceid) {
+
   if (deviceid < -1 || deviceid > RAI_MAX_DEVICE_ID) {
-    return 0;
+    return REDISMODULE_ERR;
   }
 
-  size_t key = deviceToKey(device, deviceid);
+  const size_t queue_number = deviceToKey(device, deviceid);
 
-  khiter_t key_iter = kh_get_run_queues(run_queues, key);
-  if (key_iter != kh_end(run_queues)) {
-    return 1;
+  if (queue_number<0) {
+        return REDISMODULE_ERR;
   }
 
-  RunQueueInfo* run_queue_info = RedisModule_Calloc(1, sizeof(RunQueueInfo));
-  // run_queue->run_queue_mutex = PTHREAD_MUTEX_INITIALIZER;
-  run_queue_info->run_queue = queueCreate();
+  while (queue_number >= array_len(run_queues)){
+    RunQueueInfo *run_queue_info = RedisModule_Alloc(sizeof(RunQueueInfo));
+    run_queue_info->run_queue = queueCreate();
+    pthread_cond_init(&run_queue_info->queue_condition_var, NULL);
+    pthread_mutex_init(&run_queue_info->run_queue_mutex, NULL);
+    pthread_t tid;
 
-  pthread_t tid;
-
-  if (pthread_create(&tid, NULL, RedisAI_Run_ThreadMain, run_queue_info) != 0) {
-    freeRunQueueInfo(run_queue_info);
-    return REDISMODULE_ERR;
+    if (pthread_create(&tid, NULL, RedisAI_Run_ThreadMain, run_queue_info) != 0){
+      freeRunQueueInfo(run_queue_info);
+      return REDISMODULE_ERR;
+    }
+    run_queue_info->run_thread_id = tid;
+    run_queues = array_append(run_queues, run_queue_info);
   }
```

Comment:
Here we are filling up threads for indices < queue_number that do not necessarily correspond to devices. 
I'd go with something like
```
if (queue_number >= array_len(run_queues)) {
  run_queues = array_append(run_queues, NULL);
}
RunQueueInfo *run_queue_info = RedisModule_Alloc(sizeof(RunQueueInfo));
run_queue_info->run_queue = queueCreate();
pthread_cond_init(&run_queue_info->queue_condition_var, NULL);
pthread_mutex_init(&run_queue_info->run_queue_mutex, NULL);
pthread_t tid;
if (pthread_create(&tid, NULL, RedisAI_Run_ThreadMain, run_queue_info) != 0){
  freeRunQueueInfo(run_queue_info);
  return REDISMODULE_ERR;
}
run_queue_info->run_thread_id = tid;
run_queues[queue_number] = run_queue_info;
```
I would also decrease RAI_MAX_DEVICE_ID to 32